### PR TITLE
Split nn.MultiHeadAttention into Module + functional

### PR DESCRIPTION
**Pull request by jamarshon**
Moving functions from torch/nn/modules/activation.py to torch/nn/functional.py. For functions not implemented (_get_input_buffer and _set_input_buffer), a TODO is added.

### CODE REVIEW

File: torch/nn/functional.py [hunk 1]
```diff
@@ -3077,3 +3077,249 @@ def _pad_circular(input, padding):
         input = torch.cat([input[:, :, :, :, -(padding[-5] + padding[-6]):-padding[-5]], input], dim=4)
 
     return input
+
+@weak_script
+def _in_proj(input, weight, bias, start=0, end=None):
```
Comment (cpuhrsch) [inline torch/nn/functional.py]: These should not be in functional.py scope. If you need them, move them under "multi_head_attention_forward".

Comment (jamarshon) [reply torch/nn/functional.py]: Done. also removed TODO as functions are removed in https://github.com/pytorch/pytorch/pull/20177

File: torch/nn/functional.py [hunk 2]
```diff
@@ -3077,3 +3077,195 @@ def _pad_circular(input, padding):
...
+    Args:
+        query, key, value: map a query and a set of key-value pairs to an output.
+            See "Attention Is All You Need" for more details.
+        embed_dim_to_check: total dimension of the model.
+        num_heads: parallel attention heads.
+        in_proj_weight, in_proj_bias: input projection weight and bias.
+        bias_k, bias_v: bias of the key and value sequences to be added at dim=0.
+        add_zero_attn: add a new batch of zeros to the key and
+                       value sequences at dim=1.
+        head_dim: dimension of attention heads.
```
Comment (zhangguanheng66) [inline torch/nn/functional.py]: head_dim is equal to embed_dim/num_heads. So you need to pass only two of the three parameters.

Comment (jamarshon) [reply torch/nn/functional.py]: Done.

**Review by cpuhrsch — APPROVED** [review]
Once all tests pass this should be good to. cc @zhangguanheng66

File: torch/nn/functional.py [hunk 3]
```diff
@@ -3077,3 +3077,197 @@ def _pad_circular(input, padding):
...
     return input
+
+
+@weak_script
+def multi_head_attention_forward(
+        query, key, value, embed_dim_to_check, num_heads,
+        in_proj_weight, in_proj_bias, bias_k, bias_v, add_zero_attn,
+        dropout_p, out_proj, training=True,
+        key_padding_mask=None, need_weights=True, attn_mask=None):
+    # type: (Tensor, Tensor, Tensor, int, int, Tensor, Tensor, Tensor, Tensor, bool, float, Tensor, bool, Optional[Tensor], bool, Optional[Tensor]) -> Tuple[Tensor, Tensor]
```
Comment (cpuhrsch) [inline torch/nn/functional.py]: This shows up [in the CI](https://travis-ci.org/pytorch/pytorch/jobs/531948682#L234) as too long.

**Review by zhangguanheng66 — APPROVED** [review]

File: torch/nn/functional.py [hunk 4]
```diff
@@ -3077,3 +3077,195 @@ def _pad_circular(input, padding):
...
+    def _in_proj(input, weight, bias, start=0, end=None):
+        # type: (Tensor, Tensor, Optional[Tensor], int, Optional[int]) -> Tensor
+        r"""
+        Applies a linear transformation to the incoming data: :math:`y = xA^T + b`.
+        The given integers :attr:`start` and :attr:`end` determine the subset/slice
+        of :attr:`weight` and :attr:`bias` which effectively limits the number of
+        out_features. Assumes :attr:`start` and :attr:`end` are within the bounds
+        of :math:`[0, out\_features]`.
+
+        Shape:
```
[the remark this answers is not in the document]
Comment (cpuhrsch) [reply torch/nn/functional.py]: Yes, good point!

[the remark this answers is not in the document]
Comment (jamarshon) [reply torch/nn/functional.py]: Done.

File: torch/nn/functional.py [hunk 5]
```diff
@@ -3077,3 +3077,195 @@ def _pad_circular(input, padding):
...
+        embed_dim_to_check: total dimension of the model.
+        num_heads: parallel attention heads.
+        in_proj_weight, in_proj_bias: input projection weight and bias.
+        bias_k, bias_v: bias of the key and value sequences to be added at dim=0.
+        add_zero_attn: add a new batch of zeros to the key and
+                       value sequences at dim=1.
+        head_dim: dimension of attention heads.
+        scaling: scaling factor for q.
+        dropout_p: probability of an element to be zeroed.
+        out_proj: the output projection.
```
[the remark this answers is not in the document]
Comment (jamarshon) [reply torch/nn/functional.py]: Will put in next PR.

**Comment by cpuhrsch** [comment]
Current test failures are due to temporary CI outages and restricted to AMD and Windows only.

**Comment by zhangguanheng66** [comment]
@ezyang

**Pull request closed without merging**